anvil/src/xwayland/mod.rs

```diff
@@ -0,0 +1,169 @@
+use std::{convert::TryFrom, os::unix::net::UnixStream, rc::Rc};
+
+use smithay::{
+    reexports::{
+        calloop::LoopHandle,
+        wayland_server::Client,
+    },
+    xwayland::XWindowManager,
+};
+
+use x11rb::{
+    connection::Connection as _,
+    errors::ReplyOrIdError,
+    protocol::{
+        composite::{ConnectionExt as _, Redirect},
+        xproto::{
+            ChangeWindowAttributesAux, ConfigWindow, ConfigureWindowAux, ConnectionExt as _, EventMask,
+            WindowClass,
+        },
+        Event,
+    },
+    rust_connection::{DefaultStream, RustConnection},
+};
+
+use crate::AnvilState;
+
+use x11rb_event_source::X11Source;
+
+mod x11rb_event_source;
+
+/// Implementation of [`smithay::xwayland::XWindowManager`] that is used for starting XWayland.
+/// After XWayland was started, the actual state is kept in `X11State`.
+pub struct XWm {
+    handle: LoopHandle<AnvilState>,
+    log: slog::Logger,
+}
+
+impl XWm {
+    pub fn new(handle: LoopHandle<AnvilState>, log: slog::Logger) -> Self {
+        Self { handle, log }
+    }
+}
+
+impl XWindowManager for XWm {
+    fn xwayland_ready(&mut self, connection: UnixStream, _client: Client) {
+        let (mut wm, source) = X11State::start_wm(connection, self.log.clone()).unwrap();
+        self.handle
+            .insert_source(source, move |events, _, _| {
+                for event in events.into_iter() {
+                    wm.handle_event(event)?;
+                }
+                Ok(())
+            })
+            .unwrap();
+    }
+
+    fn xwayland_exited(&mut self) {}
```

Comment:
The `XWindowManager` trait and the general XWayland api in Smithay has several issues (see https://github.com/Smithay/smithay/issues/245 ), so it'll need to be rethought and the question of how to handle restarting XWayland is a core subject. It will not be possible to handle this properly without this redesign.

Reply:
> Or is a closed fd automatically removed?

I'm not even sure if the FD is closed when the X11 connection breaks. The `UnixConnection` is owned by `x11rb::rust_connection::RustConnection`, which lives in an `Rc` that is inside the `x11rb_event_source::X11Source` which is given to calloop. The same `Rc` is also referenced via `Client::data_map()` for the X11 client. The `Client` in turn is referenced by the callback used for handling X11 events. It is very well possible that this ends up being a cyclic reference where everything keeps everything alive and the `UnixConnection` is not dropped.

Well, actually, when the X11 connection is closed, the `UnixConnection` becomes readable and thus the `X11Source` wakes up. This does `conn.poll_for_event()`, gets some kind of error (`x11rb::ConnectionError::IOError`?), wraps this error in an `IOError` and returns it from `X11Source::process_events`. I do not know what calloop then does with this error...

Looks like calloop returns the error from `event_loop.dispatch()`. The code in `anvil/src/winit.rs` and `udev.rs` just checks for `is_err()` and then silently exits without any helpful message.

Since smithay's XWayland code does `None.unwrap()` when Xwayland dies, I haven't done much in this direction expect thinking "let's hope this somehow works". The above is more than I ever did before.